pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -69,20 +108,40 @@ func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTime
 				}
 			}
 		}()
+
+		go func() {
+			for {
+				select {
+				case <-p.close:
+					return
+				case <-time.After(time.Minute):
+					// We copy the state of blocks that are being lazy loaded from LazyBinaryReaders using lock.
+					p.lazyReadersMx.Lock()
+					p.lazyLoadedTracker.copyLazyLoadedState(p.lazyReaders)
+					p.lazyReadersMx.Unlock()
+
+					// Then we persist the state to files so that we are not holding lock for too long.
+					if err := p.lazyLoadedTracker.persist(); err != nil {
+						level.Warn(p.logger).Log("msg", "failed to persist list of lazy-loaded index headers", "err", err)
+					}
```

Comment:
I find this pattern very confusing.

The pool requires injecting an entity called _HeadersLazyLoadedTracker_ which is a combination of configuration (Path) and half-loaded state (`State` but only has `userID`). Then we call a `copyLazyLoadedState` method that mutates the _tracker_ (not the state, suprisingly), and then we ask the tracker to persist the state.

Is this state going to be read from the `ReaderPool` later or from the caller of `ReaderPool`? 

Why do we need to inject the `HeadersLazyLoadedTracker` into `NewReaderPool` alongside with _some tracker config_ (enabled? idle timeout, hardcoded 1 minute persist interval)? Can we just inject the `path` and `user`, and let the `ReaderPool` build its tracker?

What I expect from the code in it's current form:
```go
					// lazyLoadedState takes care of holding the mutex, and it returns the map of blockID to last used millisecond timestamp.
					state := p.lazyLoadedState()

					// lazyLoadedTracker persists the provided state, without mutating itself.
					if err := p.lazyLoadedTracker.persist(state); err != nil {
						level.Warn(p.logger).Log("msg", "failed to persist list of lazy-loaded index headers", "err", err)
					}
```

This however doesn't answer the question of why we need to inject the tracker from outside, creating a dependency here that isn't needed for correct pool function.

What I would do, is implementing a method:

```go
// LoadedBlocks returns the map of lazy-loaded block IDs and the last time they were used in milliseconds.
func (p *ReaderPool) LoadedBlocks() map[string]int64 {
	p.lazyReadersMx.Lock()
	defer p.lazyReadersMx.Unlock()

	blocks := make(map[string]int64, len(p.lazyReaders))
	for r := range p.lazyReaders {
		blocks[r.blockID] = r.usedAt.Load() / int64(time.Millisecond)
	}

	return blocks
}
```

So we could make the tracker depend on the lazy loaded pool (instead of having the pool depending on the tracker). Then the tracker can take it's own configuration, a dependency `interface { LoadedBlocks() map[string]int64 }` and keep it's own ticker. This would also make testing much easier (as it's straitforward to test `LoadedBlocks` and it's easy to mock the interface)